### PR TITLE
Rename various settings

### PR DESCRIPTION
**Pull request by Colengms**
|Old name|New name|
|----------------|-------------------------------|
|C_Cpp.clang_format_style|C_Cpp.clangFormat.style|
|C_Cpp.clang_format_path|C_Cpp.clangFormat.path|
|C_Cpp.clang_format_fallbackStyle|C_Cpp.clangFormat.fallbackStyle|
|C_Cpp.clang_format_sortIncludes|C_Cpp.clangFormat.sortIncludes|
|C_Cpp.autocomplete|C_Cpp.intelliSense.autocomplete|
|C_Cpp.errorSquiggles|C_Cpp.intelliSense.errorSquiggles|
|C_Cpp.exclusionPolicy|C_Cpp.workspaceIndexing.exclusionPolicy|
|C_Cpp.formatting|C_Cpp.formatting.enabled|
|C_Cpp.intelliSenseCachePath|C_Cpp.intelliSense.cache.path|
|C_Cpp.intelliSenseCacheSize|C_Cpp.intelliSense.cache.size|
|C_Cpp.intelliSenseEngine|C_Cpp.intelliSense.engine|
|C_Cpp.intelliSenseEngineFallback|C_Cpp.intelliSense.fallback|
|C_Cpp.workspaceParsingPriority|C_Cpp.workspaceIndexing.priority|

Added: C_Cpp.intelliSense.enabled
Removed: C_Cpp.navigation.length
Removed: clang_format_formatOnSave

There is a native side change as well, to use the new settings and to convert workspace_didChangeConfiguration to our own cpptools_didChangeSettings.

Removed old AB testing.

### CODE REVIEW

File: Extension/src/LanguageServer/settings.ts [hunk 1]
```diff
@@ -32,35 +47,56 @@ export class CppSettings extends Settings {
...
+            result = info.workspaceFolderValue;
+        } else if (info.workspaceValue) {
+            result = info.workspaceValue;
+        } else if (info.globalValue) {
+            result = info.globalValue;
+        } else {
+            let value: string = this.Section.get<string>("formatting");
+            if (value === "Default" || value === "Disabled") {
+                return value;
+            }
```
Comment (bobbrow) [inline Extension/src/LanguageServer/settings.ts:78]: I think you need an `else` here that sets `result` to the default value or else we'll end up disabling formatting on line 76.

Or can you just call `getWithFallback` and translate the boolean result to Default | Disabled?  It seems like this function duplicates a lot of that code.  IOW, would this work?
```ts
    public get formatting(): string {
        return this.getWithFallback<boolean>("formatting.enabled", "formatting") ? "Default" : "Disabled";
    }
```

File: Extension/src/main.ts [hunk 2]
```diff
@@ -288,59 +288,36 @@ async function postInstall(info: PlatformInformation): Promise<void> {
...
             reloadMessageShown = true;
             util.promptForReloadWindowDueToSettingsChange();
         }
     }));
     getTemporaryCommandRegistrarInstance().activateLanguageServer();
 
-    // Update default for C_Cpp.intelliSenseEngine based on A/B testing settings.
-    // (this may result in rewriting the package.json file)
-
-    let abTestSettings: cpptoolsJsonUtils.ABTestSettings = cpptoolsJsonUtils.getABTestSettings();
```
Comment (bobbrow) [inline Extension/src/main.ts:313]: Can you remove the corresponding deprecated properties from `./cpptools.json` as well?

**Review by bobbrow — COMMENTED** [review]
Adding some comments.

File: Extension/package.json [hunk 3]
```diff
@@ -234,7 +235,7 @@
           "description": "%c_cpp.configuration.autoAddFileAssociations.description%",
           "scope": "resource"
         },
-        "C_Cpp.workspaceParsingPriority": {
+        "C_Cpp.workspaceIndexing.priority": {
```
Comment (sean-mcmanus) [inline Extension/package.json]: This really does apply to the "parsing" and not the "indexing". Was there some other reason for this change?

File: Extension/package.json [hunk 4]
```diff
@@ -315,13 +316,13 @@
           "description": "%c_cpp.configuration.configurationWarnings.description%",
           "scope": "resource"
         },
-        "C_Cpp.intelliSenseCachePath": {
+        "C_Cpp.intelliSense.cachePath": {
           "type": "string",
```
Comment (sean-mcmanus) [inline Extension/package.json:320]: Should this be cache.path and cache.size?

File: Extension/src/LanguageServer/client.ts [hunk 5]
```diff
@@ -961,6 +962,32 @@ export class DefaultClient implements Client {
     public onDidChangeSettings(event: vscode.ConfigurationChangeEvent): { [key: string] : string } {
         let changedSettings: { [key: string] : string } = this.settingsTracker.getChangedSettings();
         this.notifyWhenReady(() => {
+            // Unlike the LSP message, the event does not contain all settings as a payload, so we need to
+            // build a new JSON object with everything we need on the native side.
+            let settings: any = {
+                "C_Cpp": { ...vscode.workspace.getConfiguration("C_Cpp", this.RootUri) },
+                "files": vscode.workspace.getConfiguration("files", this.RootUri),
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/client.ts]: Only the files.exclude, search.exclude, and editor.tabSize (missing) settings are needed.

File: Extension/package.json (hunk 4 above)
Comment (Colengms) [reply Extension/package.json:320]: It was previously cachePath and cacheSize in the OneNote page, and I had split them.  But, then I encountered a complication in SettingsTracker which would not be an issue with cachePath and cacheSize.  Does anyone feel strongly this should be one way or the other?

File: Extension/src/LanguageServer/client.ts (hunk 5 above)
Comment (Colengms) [reply Extension/src/LanguageServer/client.ts]: Where is editor.tabSize referenced?  In cpptools_didChangeSettings (was workspace_didChangeConfiguration) where the settings json is referenced, it's to get "C_Cpp", "files" and "search".

File: Extension/src/LanguageServer/settings.ts (hunk 1 above)
Comment (Colengms) [reply Extension/src/LanguageServer/settings.ts:78]: Fixed.  As discussed, I'm accounting for settings under "formatting" potentially colliding with the deprecated "formatting" element.

File: Extension/src/LanguageServer/settings.ts [hunk 6]
```diff
@@ -32,35 +47,84 @@ export class CppSettings extends Settings {
...
+            // intelliSenseEngine may be set to disabled, but that is reflected in intelliSense.enabled now.
+            result = "Default";
+        }
+        return result;
+    }
+    public get intelliSenseEnabled(): boolean
+    {
+        let result: boolean;
+        let info = this.Section.inspect<boolean>("intelliSense.enabled");
+        if (info.workspaceFolderValue != undefined) {
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/settings.ts]: Should this be !==?

File: Extension/src/LanguageServer/client.ts [hunk 7]
```diff
@@ -891,38 +890,49 @@ export class DefaultClient implements Client {
...
             initializationOptions: {
-                clang_format_path: util.resolveVariables(settings.clangFormatPath, this.AdditionalEnvironment),
-                clang_format_style: settings.clangFormatStyle,
-                clang_format_fallbackStyle: settings.clangFormatFallbackStyle,
-                clang_format_sortIncludes: settings.clangFormatSortIncludes,
-                formatting: settings.formatting,
+                clangFormat: {
+                    path: util.resolveVariables(settings.clangFormatPath, this.AdditionalEnvironment),
+                    style: settings.clangFormatStyle,
+                    fallbackStule: settings.clangFormatFallbackStyle,
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/client.ts]: fallbackStule->fallbackStyle

**Review by sean-mcmanus — APPROVED** [review]

**Pull request merged**